### PR TITLE
Support POST requests in React Native store

### PR DESCRIPTION
**Pull request by fluiddot**
**Related PRs:**
* https://github.com/WordPress/gutenberg/pull/49371
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5596
* Android
    * https://github.com/wordpress-mobile/WordPress-Android/pull/18181
    * https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691
* iOS
    * https://github.com/wordpress-mobile/WordPressKit-iOS/pull/589
    * https://github.com/wordpress-mobile/WordPress-iOS/pull/20428

This PR introduces the support for POST requests in the React Native store. With this change, the Gutenberg editor will be able to make POST requests.

The main changes of this PR are:
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691/commits/737110e34409e22db4cff7d346829614f9f5de37: With the introduction of POST requests, I renamed the functions to differentiate GET and POST request functions in the React Native store.
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691/commits/3122c6a9ca7183f17f1eabe7c576aff572a6e91b: Add request methods enum and update the GET functions to run only when the method passed is `GET`.
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691/commits/69faa29afe2138126221674124ee18d7670a199d: Add functions to handle POST requests.
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691/commits/0ccc8ed4880213826949c27ce7f4e2e39dcc08dc: Add unit tests for new POST request functions.
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691/commits/677ed381c684aa0c368debfc4d25f015969d0687: Originally the `body` parameter passed in POST requests had type `Map<String, String>`. However, I noticed that the `GsonRequest` constructor accepts any object 
 for the `body` parameter not just `String`. In fact, it uses the type `Map<String, Object>`. This commit updates the types in `WPAPIGsonRequestBuilder` of that parameter, but since it's a Kotlin file, we use `Map<String, Any>`.
- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2691/commits/267fce198b509e4db5ab1eab6cf97773dfeae988: Following the above commit, update the `body` type to `Map<String, Any>` in POST requests functions of React Native store.

**NOTE:** I haven't added tests against real sites (i.e. tests with file names `ReleaseStack_*`). I'm planning to do that in a separate PR.

### Testing Details

**The Gutenberg editor doesn't make POST requests yet**, hence this change can only be tested with a test PR: https://github.com/wordpress-mobile/WordPress-Android/pull/18202.

### CODE REVIEW

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/ReactNativeStore.kt [hunk 1]
```diff
@@ -196,6 +263,7 @@ class ReactNativeStore @VisibleForTesting constructor(
                 val url = path?.let {
                     val newPath = it
                             .replace("wp/v2".toRegex(), "wp/v2/sites/$wpComSiteId")
+                            .replace("wpcom/v2".toRegex(), "wpcom/v2/sites/$wpComSiteId")
```
Comment (fluiddot) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/ReactNativeStore.kt:266]: Adding a new namespace that will be required to make requests related to VideoPress ([example](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5528)).

**Review by SiobhyB — APPROVED** [review]
Approved via https://github.com/wordpress-mobile/WordPress-Android/pull/18181#pullrequestreview-1365464022.

As mentioned in that review, it could be a good idea to get a second review from a native developer, as I'm certainly no expert in Kotlin or Java. From my current knowledge, though, the code matches the patterns I'd expect to see and functionally passes the manual tests too.

Thanks again for all your work on these PRs, Carlos! 👏 👏 🎉

**Comment by fluiddot** [comment]
> Approved via [wordpress-mobile/WordPress-Android#18181 (review)](https://github.com/wordpress-mobile/WordPress-Android/pull/18181#pullrequestreview-1365464022).

Thanks @SiobhyB for reviewing the PR 🙇 !

> As mentioned in that review, it could be a good idea to get a second review from a native developer, as I'm certainly no expert in Kotlin or Java. From my current knowledge, though, the code matches the patterns I'd expect to see and functionally passes the manual tests too.

Yep, good point. I also would like to have a second review from a native developer. I'm checking the history of some of the files and noticed that @ParaskP7 was one of the last contributors. @ParaskP7 let me know if you could take a quick look at the PR to confirm that the changes make sense from a native POV. Thanks for the help 🙇 !

**Comment by ParaskP7** [comment]
👋 @fluiddot and thanks for the ping!

> I'm checking the history of some of the files and noticed that @ParaskP7 was one of the last contributors.

Unfortunately this is a bit misleading Carlos, you seeing me being the last contributor on those files just because of `Analysis` related commits there, which have nothing to do with the functionality itself.

PS: I was doing as part of improving static analysis and the tooling around these files, probably while integrating `Detekt` into FluxC.

> @ParaskP7 let me know if you could take a quick look at the PR to confirm that the changes make sense from a native POV. Thanks for the help 🙇 !

Having said the above, I think it is best for someone else to take a look at the PR as I have very little-to-no knowledge on those file and wouldn't want to mislead you with my review on anything `React Native` related.

> Yep, good point. I also would like to have a second review from a native developer.

Looking at the history of those files, I would recommend either @planarvoid , @mchowning or @antonis (from the looks of it). 🤞 

-----

Apologies for ping-ponging you like that, but if you still want me to review those I could definitely take a look next week (on Fridays I am very busy with release management related duties). 🙏 🤔 🙏

**Comment by fluiddot** [comment]
> Unfortunately this is a bit misleading Carlos, you seeing me being the last contributor on those files just because of `Analysis` related commits there, which have nothing to do with the functionality itself.
> 
> PS: I was doing as part of improving static analysis and the tooling around these files, probably while integrating `Detekt` into FluxC.

Oh, good point, sorry for the ping @ParaskP7. I'll remove you from the reviewers list 💨 .

> Having said the above, I think it is best for someone else to take a look at the PR as I have very little-to-no knowledge on those file and wouldn't want to mislead you with my review on anything React Native related.

Of course, makes total sense to me.

> Looking at the history of those files, I would recommend either @planarvoid , @mchowning or @antonis (from the looks of it). 🤞

Thanks for the recommendation, I've also asked internally for help in case other Android developers could take a look. Thanks for the help 🙇 !

**Review by ovitrif — APPROVED** [review]
I've reviewed the code carefully and tested with a WPCOM site using the test PR on WPAndroid.

It works as expected and the code looks good to me, great work in all aspects @fluiddot 🙇🏻 🥇 :
- refactoring changes make sense to me
- extending the code to support post request is clean
- it's easy to distinguish the 2 request methods in code
- unit tests are on point

:shipit:

**Review by thomashorta — COMMENTED** [review]
The code looks excellent! :shipit: 

Unfortunately, I was not able to test it but I see Ovi and others tested it!

**Pull request merged**